BLS12-381 host functions (#156)
polkadot-fellows/RFCs: text/0156-bls-signatures.md

# RFC-0156: Add BLS12-381 Host Functions

|                 |                                                                                             |
| --------------- | ------------------------------------------------------------------------------------------- |
| **Start Date**  | 2025-10-15                                                                                  |
| **Description** | Introduce BLS12-381 host functions                                                          |
| **Author**      | Someone Unknown                                                                             |

## Summary

> Introduce new host functions allowing runtimes to generate BLS12-381 keys, signatures, and proofs of possession.

Reviewer: ```suggestion
Introduce new host functions allowing runtimes to generate BLS12-381 keys, signatures, and proofs of possession. Additionally it allows runtime to generate paired (ECDSA,BLS12-381) keys to enable BEEFY to support both signatures at the same time.
```

* `out` is a pointer ([Definition 215](https://polkadotspec.dev/chap-host-api#defn-runtime-pointer)) to an output buffer, 144 bytes long, where the public key will be written.

### ext_crypto_bls381_generate_proof_of_possession

Generates a BLS12-381 Proof Of Possession for a given public key and owner identifier.

#### Prototype

```wat
(func $ext_crypto_bls381_generate_proof_of_possession_version_1
 (param $id i32) (param $pub_key i32) (param $out i32) (result i64))
```

#### Arguments

* `id` is a pointer ([Definition 215](https://polkadotspec.dev/chap-host-api#defn-runtime-pointer)) to a key type identifier ([Definition 220](https://polkadotspec.dev/chap-host-api#defn-key-type-id)). The function will panic if the identifier is invalid;
* `pub_key` is a pointer ([Definition 215](https://polkadotspec.dev/chap-host-api#defn-runtime-pointer)) to a public key, 144 bytes long;

> * `out` is a pointer ([Definition 215](https://polkadotspec.dev/chap-host-api#defn-runtime-pointer)) to an output buffer, 224 bytes long, where the proof of possession will be written.

Reviewer: ```suggestion
* `out` is a pointer ([Definition 215](https://polkadotspec.dev/chap-host-api#defn-runtime-pointer)) to an output buffer, 112 bytes long, where the proof of possession will be written.
```

### ext_crypto_bls381_sign

Signs an input message using a given BLS12-381 key.

#### Prototype

```wat
(func $ext_crypto_bls381_sign_version_1
 (param $id i32) (param $pub_key i32) (param $msg i64) (param $out i64) (result i64))
```

#### Arguments

* `id` is a pointer ([Definition 215](https://polkadotspec.dev/chap-host-api#defn-runtime-pointer)) to a key type identifier ([Definition 220](https://polkadotspec.dev/chap-host-api#defn-key-type-id)). The function will panic if the identifier is invalid;
* `pub_key` is a pointer ([Definition 215](https://polkadotspec.dev/chap-host-api#defn-runtime-pointer)) to public key bytes (as returned by `ext_crypto_bls381_public_keys` function);
* `msg` is a pointer-size ([Definition 216](https://polkadotspec.dev/chap-host-api#defn-runtime-pointer-size)) to a message that is to be signed;
* `out` is a pointer ([Definition 215](https://polkadotspec.dev/chap-host-api#defn-runtime-pointer)) to an output buffer, 112 bytes long, where the signature will be written.

#### Result

The function returns `0` on success. On error, `-1` is returned, and the output buffer should be considered uninitialized.

### ext_crypto_ecdsa_bls381_generate

> Generates a combination ECDSA & BLS12-381 key for a given key type using an optional `seed`, stores it in the keystore, and returns the corresponding public key.

Reviewer: ```suggestion
Generates a pair of (ECDSA,BLS12-381) keys for a given key type using an optional `seed`, stores it in the keystore, and returns the corresponding public key. The keystore and the runtime regard this pair as a single key.
```

#### Prototype

```wat
(func $ext_crypto_ecdsa_bls381_generate_version_1
 (param $id i32) (param $seed i64) (param $out i32))
```

#### Arguments

* `id` is a pointer ([Definition 215](https://polkadotspec.dev/chap-host-api#defn-runtime-pointer)) to a key type identifier ([Definition 220](https://polkadotspec.dev/chap-host-api#defn-key-type-id)). The function will panic if the identifier is invalid;
* `seed` is a pointer-size ([Definition 216](https://polkadotspec.dev/chap-host-api#defn-runtime-pointer-size)) to a SCALE-encoded Option value ([Definition 200](https://polkadotspec.dev/id-cryptography-encoding#defn-option-type)) containing a BIP-39 seed which must be valid UTF-8. The function will panic if the seed is not a valid UTF-8;
* `out` is a pointer ([Definition 215](https://polkadotspec.dev/chap-host-api#defn-runtime-pointer)) to an output buffer, 177 bytes long, where the public key will be written.

### ext_crypto_ecdsa_bls381_public_keys

> Retrieves all ECDSA & BLS12-381 public keys of a given type from the keystore.

Reviewer: ```suggestion
Retrieves all paired (ECDSA,BLS12-381) public keys of a given type from the keystore.
```

#### Prototype

```wat
(func $ext_crypto_ecdsa_bls381_public_keys_version_1
 (param $id i32) (param $out i64) (result i32))
```

#### Arguments

* `id` is a pointer ([Definition 215](https://polkadotspec.dev/chap-host-api#defn-runtime-pointer)) to a key type identifier ([Definition 220](https://polkadotspec.dev/chap-host-api#defn-key-type-id)). The function will panic if the identifier is invalid;
* `out` is a pointer-size ([Definition 216](https://polkadotspec.dev/chap-host-api#defn-runtime-pointer-size)) to a buffer where the public keys of the given type known to the keystore will be stored consecutively. The value is actually stored only if the buffer is large enough. Otherwise, the buffer is not written into, and its contents are unchanged.

#### Result

> The result is an unsigned 32-bit integer representing the total size in bytes required to store all public keys of the given type. The number of keys can be determined by dividing this value by the known key size (177 bytes for ECDSA & BLS12-381). A value of `0` indicates that no keys of the given type are known to the keystore.

Reviewer: ```suggestion
The result is an unsigned 32-bit integer representing the total size in bytes required to store all public keys of the given type. The number of keys can be determined by dividing this value by the known key size (177 bytes for (ECDSA,BLS12-381)). A value of `0` indicates that no keys of the given type are known to the keystore.
```

#### Prototype

```wat
(func $ext_crypto_bls381_generate_version_1
 (param $id i32) (param $seed i64) (param $out i32))
```

#### Arguments

* `id` is a pointer ([Definition 215](https://polkadotspec.dev/chap-host-api#defn-runtime-pointer)) to a key type identifier ([Definition 220](https://polkadotspec.dev/chap-host-api#defn-key-type-id)). The function will panic if the identifier is invalid;
* `seed` is a pointer-size ([Definition 216](https://polkadotspec.dev/chap-host-api#defn-runtime-pointer-size)) to a SCALE-encoded Option value ([Definition 200](https://polkadotspec.dev/id-cryptography-encoding#defn-option-type)) containing a BIP-39 seed which must be valid UTF-8. The function will panic if the seed is not a valid UTF-8;
* `out` is a pointer ([Definition 215](https://polkadotspec.dev/chap-host-api#defn-runtime-pointer)) to an output buffer, 144 bytes long, where the public key will be written.

### ext_crypto_bls381_generate_proof_of_possession

Generates a BLS12-381 Proof Of Possession for a given public key and owner identifier.

#### Prototype

```wat

> (func $ext_crypto_bls381_generate_proof_of_possession_version_1

Author: I'll be waiting for input on this one, then. CC @P1sar
